<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_22) on Mon Jun 28 16:44:44 EDT 2010 -->
<TITLE>
Uses of Class gffParser.MRNA
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Class gffParser.MRNA";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../gffParser/MRNA.html" title="class in gffParser"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?gffParser/class-use/MRNA.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MRNA.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>gffParser.MRNA</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#genomeBrowser"><B>genomeBrowser</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#gffParser"><B>gffParser</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="genomeBrowser"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A> in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A> declared as <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></CODE></FONT></TD>
<TD><CODE><B>UpdateAnimation.</B><B><A HREF="../../genomeBrowser/UpdateAnimation.html#isoform">isoform</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A> with type parameters of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/List.html" title="class or interface in java.util">List</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#currentlyViewingIsoforms">currentlyViewingIsoforms</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A> with parameters of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#animatedLoadShortReads(java.util.ArrayList, gffParser.MRNA, int)">animatedLoadShortReads</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;listOfSamRecords,
                       <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                       int&nbsp;iTotalNumberOfReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads Short Read data but also animates the exons so that they float in the direction of the 
 new short reads</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#fillInIsoformSpecificDensityPlot(java.util.ArrayList, gffParser.MRNA)">fillInIsoformSpecificDensityPlot</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iShortReads,
                                 <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;iMRNA)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#findCompatibleShortReads(gffParser.MRNA, java.util.ArrayList)">findCompatibleShortReads</A></B>(<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                         <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Find compatible short reads.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#getCompatibleSamRecords(gffParser.MRNA, java.util.ArrayList)">getCompatibleSamRecords</A></B>(<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                        <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../../genomeBrowser/ShortRead.html" title="class in genomeBrowser">ShortRead</A>&gt;</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#getCompatibleShortReads(gffParser.MRNA, java.util.ArrayList)">getCompatibleShortReads</A></B>(<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                        <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a list of compatible SAMRecords to a specific isoform and fills the Junction List, making it accessible to draw</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#loadJunctionList(gffParser.MRNA, java.util.ArrayList)">loadJunctionList</A></B>(<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                 <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Method parameters in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A> with type arguments of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#identifyConstituitiveBases(java.util.Collection)">identifyConstituitiveBases</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;mrnaList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an ArrayList of genomic coordinates which are constitutive.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#isFoundAmongAllMRNA(java.util.Collection, int)">isFoundAmongAllMRNA</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;mrnaList,
                    int&nbsp;absolutePosition)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Checks if position is found among all isoforms.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#loadArrayOfUnweightedIsoforms(java.util.List)">loadArrayOfUnweightedIsoforms</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/List.html" title="class or interface in java.util">List</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;listOfMRNA)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Load array of unweighted isoforms without changing the parameters</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#loadArrayOfWeightedIsoforms(java.util.List)">loadArrayOfWeightedIsoforms</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/List.html" title="class or interface in java.util">List</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;listOfMRNA)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Load array of weighted isoforms without changing the parameters</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#loadNewArrayOfUnweightedIsoforms(java.util.Collection, int, int, boolean)">loadNewArrayOfUnweightedIsoforms</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;isoforms,
                                 int&nbsp;iAbsoluteStartOfGene,
                                 int&nbsp;iAbsoluteLengthOfGene,
                                 boolean&nbsp;strand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads information for positioning unweighted isoforms into
 global arrays where they will be drawn through the draw function.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProcessingApplet.</B><B><A HREF="../../genomeBrowser/ProcessingApplet.html#loadNewArrayOfWeightedIsoforms(java.util.List, int, int, boolean)">loadNewArrayOfWeightedIsoforms</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/List.html" title="class or interface in java.util">List</A>&lt;<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;listOfMRNA,
                               int&nbsp;iAbsoluteStartOfGene,
                               int&nbsp;iAbsoluteLengthOfGene,
                               boolean&nbsp;strand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Loads information for positioning Weighted isoforms into 
 global arrays.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../genomeBrowser/package-summary.html">genomeBrowser</A> with parameters of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../genomeBrowser/UpdateAnimation.html#UpdateAnimation(genomeBrowser.ProcessingApplet, gffParser.MRNA, java.util.HashMap, java.util.List)">UpdateAnimation</A></B>(<A HREF="../../genomeBrowser/ProcessingApplet.html" title="class in genomeBrowser">ProcessingApplet</A>&nbsp;iParent,
                <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;iIsoform,
                <A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>,<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;&nbsp;iCompatibleDensityMap,
                <A HREF="http://java.sun.com/javase/6/docs/api/java/util/List.html" title="class or interface in java.util">List</A>&lt;<A HREF="../../genomeBrowser/Rectangle_Weighted.html" title="class in genomeBrowser">Rectangle_Weighted</A>&gt;&nbsp;iWeightedIsoforms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="gffParser"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A> in <A HREF="../../gffParser/package-summary.html">gffParser</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../gffParser/package-summary.html">gffParser</A> with type parameters of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) &nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/String.html" title="class or interface in java.lang">String</A>,<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;</CODE></FONT></TD>
<TD><CODE><B>Gene.</B><B><A HREF="../../gffParser/Gene.html#hashMRNA">hashMRNA</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../gffParser/package-summary.html">gffParser</A> that return types with arguments of type <A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/String.html" title="class or interface in java.lang">String</A>,<A HREF="../../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;</CODE></FONT></TD>
<TD><CODE><B>Gene.</B><B><A HREF="../../gffParser/Gene.html#getMRNA()">getMRNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../gffParser/MRNA.html" title="class in gffParser"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?gffParser/class-use/MRNA.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MRNA.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
