<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_22) on Mon Aug 16 15:05:28 EDT 2010 -->
<TITLE>
Statistics
</TITLE>

<META NAME="keywords" CONTENT="spie.Statistics class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Statistics";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Statistics.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../spie/Read.html" title="class in spie"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../spie/Statistics.Method.html" title="enum in spie"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?spie/Statistics.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Statistics.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
spie</FONT>
<BR>
Class Statistics</H2>
<PRE>
<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html" title="class or interface in java.lang">java.lang.Object</A>
  <IMG SRC="../resources/inherit.gif" ALT="extended by "><B>spie.Statistics</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>Statistics</B><DT>extends <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html" title="class or interface in java.lang">Object</A></DL>
</PRE>

<P>
The Class Statistics.
 
 This class performs a host of statistical analysis.
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Vincent Xue</DD>
</DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->

<A NAME="nested_class_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Nested Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.Method.html" title="enum in spie">Statistics.Method</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;<A HREF="../spie/Statistics.Method.html" title="enum in spie">Statistics.Method</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#method">method</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The default method to use when calculating weights.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#overhang">overhang</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The overhang used for calculating.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#readLength">readLength</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The short read length.(Default is 35)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#totalNumberOfReads">totalNumberOfReads</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The total number of reads in a BAM file.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected </CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#Statistics()">Statistics</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#convertShortReadsToSamRecords(java.util.ArrayList)">convertShortReadsToSamRecords</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;iShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convert short reads to sam records.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getAllReads_Per_KTotalPossiblePositions(int, int, java.util.ArrayList, boolean)">getAllReads_Per_KTotalPossiblePositions</A></B>(int&nbsp;absoluteStart,
                                        int&nbsp;absoluteEnd,
                                        <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads,
                                        boolean&nbsp;endExon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the ratio of all reads (junction and body) to the total number of possible start
 positions for a short read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getAverage_ReadsPerBase_PerExon(int, int, java.util.ArrayList)">getAverage_ReadsPerBase_PerExon</A></B>(int&nbsp;absoluteStart,
                                int&nbsp;absoluteEnd,
                                <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the ratio of the sum of reads per base to the total number of bases.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getBodyReads_per_KExon(int, int, java.util.ArrayList)">getBodyReads_per_KExon</A></B>(int&nbsp;absoluteStart,
                       int&nbsp;absoluteEnd,
                       <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the number of body short reads divided by the exon length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getCompatibleReads(gffParser.MRNA, java.util.ArrayList)">getCompatibleReads</A></B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                   <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets a ArrayList of compatible ShortReads.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getCompatibleSamRecords(gffParser.MRNA, java.util.ArrayList)">getCompatibleSamRecords</A></B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                        <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSamRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets a compatible ArrayList of SAMRecords
 (it calls getCompatibleShortReadsFirst).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getConstituitiveBases(java.util.Collection)">getConstituitiveBases</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;mrnaList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an ArrayList of genomic coordinates which are constitutive.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getConstitutiveIntervals(java.util.ArrayList)">getConstitutiveIntervals</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;&nbsp;inputArrayOfConstitutivePositions)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the constitutive intervals.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>,<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getDensityMap(int, int, java.util.ArrayList)">getDensityMap</A></B>(int&nbsp;absoluteStart,
              int&nbsp;absoluteEnd,
              <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the density map.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>,<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getIsoformDensityMap(int, int, gffParser.MRNA, java.util.ArrayList)">getIsoformDensityMap</A></B>(int&nbsp;absoluteStart,
                     int&nbsp;absoluteEnd,
                     <A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSamRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the isoform density map.A density map is a hash map that has all the genomic coordinates as
 keys and the number of short reads that cross it as values</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getNumberOfBodyReads(gffParser.MRNA, int, int, java.util.ArrayList)">getNumberOfBodyReads</A></B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                     int&nbsp;absoluteStart,
                     int&nbsp;absoluteEnd,
                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the number of body reads as specified by the coordinates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getNumberOfJunctionReads(gffParser.MRNA, int, int, java.util.ArrayList)">getNumberOfJunctionReads</A></B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                         int&nbsp;absoluteStart,
                         int&nbsp;absoluteEnd,
                         <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the number of junction reads as specified by the coordinates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getRPKM(java.util.ArrayList, java.util.ArrayList, int)">getRPKM</A></B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;allSAMRecord,
        <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt;&nbsp;constitutiveIntervals,
        int&nbsp;totalNumberOfReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the rPKM.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private static&nbsp;<A HREF="../spie/Read.html" title="class in spie">Read</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getShortReadIfCompatible(net.sf.samtools.SAMRecord, java.util.ArrayList, java.util.Collection)">getShortReadIfCompatible</A></B>(net.sf.samtools.SAMRecord&nbsp;samRecord,
                         <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt;&nbsp;samInterval,
                         <A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../gffParser/Exon.html" title="class in gffParser">Exon</A>&gt;&nbsp;exonList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets a new ShortRead if the SAMRecord is compatible
 
 1) First find the exon where the short read begins
 2) Catch the case of a single interval --See if the short read is completely within the exon
 3) If more than one interval, check each interval with the next ones.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getStandardDeviation_ReadsPerBase(int, int, java.util.ArrayList)">getStandardDeviation_ReadsPerBase</A></B>(int&nbsp;absoluteStart,
                                  int&nbsp;absoluteEnd,
                                  <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the standard deviation_ reads per base.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getStandardDeviation(int, int, java.util.ArrayList)">getStandardDeviation</A></B>(int&nbsp;absoluteStart,
                     int&nbsp;absoluteEnd,
                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getWeightOfExon(gffParser.Exon, java.util.ArrayList, java.lang.Boolean)">getWeightOfExon</A></B>(<A HREF="../gffParser/Exon.html" title="class in gffParser">Exon</A>&nbsp;exon,
                <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleMRNAReads,
                <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Boolean.html" title="class or interface in java.lang">Boolean</A>&nbsp;isEndExon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the weight of an exon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#getWeightOfJunction(drawableObjects.Junction)">getWeightOfJunction</A></B>(<A HREF="../drawableObjects/Junction.html" title="class in drawableObjects">Junction</A>&nbsp;junction)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the weight of junction.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#setMethod(spie.Statistics.Method)">setMethod</A></B>(<A HREF="../spie/Statistics.Method.html" title="enum in spie">Statistics.Method</A>&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the method to use for calculating weights.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#setOverhang(int)">setOverhang</A></B>(int&nbsp;iOverhang)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the overhang.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#setReadLength(int)">setReadLength</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the short read length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../spie/Statistics.html#setTotalNumberOfReads(int)">setTotalNumberOfReads</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the total number of reads.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html" title="class or interface in java.lang">Object</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#clone()" title="class or interface in java.lang">clone</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#equals(java.lang.Object)" title="class or interface in java.lang">equals</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#finalize()" title="class or interface in java.lang">finalize</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#getClass()" title="class or interface in java.lang">getClass</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#hashCode()" title="class or interface in java.lang">hashCode</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#notify()" title="class or interface in java.lang">notify</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#notifyAll()" title="class or interface in java.lang">notifyAll</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#toString()" title="class or interface in java.lang">toString</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#wait()" title="class or interface in java.lang">wait</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#wait(long)" title="class or interface in java.lang">wait</A>, <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Object.html#wait(long, int)" title="class or interface in java.lang">wait</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="totalNumberOfReads"><!-- --></A><H3>
totalNumberOfReads</H3>
<PRE>
private static int <B>totalNumberOfReads</B></PRE>
<DL>
<DD>The total number of reads in a BAM file.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="overhang"><!-- --></A><H3>
overhang</H3>
<PRE>
private static int <B>overhang</B></PRE>
<DL>
<DD>The overhang used for calculating. (Default is 4)
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="method"><!-- --></A><H3>
method</H3>
<PRE>
private static <A HREF="../spie/Statistics.Method.html" title="enum in spie">Statistics.Method</A> <B>method</B></PRE>
<DL>
<DD>The default method to use when calculating weights.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="readLength"><!-- --></A><H3>
readLength</H3>
<PRE>
private static int <B>readLength</B></PRE>
<DL>
<DD>The short read length.(Default is 35)
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Statistics()"><!-- --></A><H3>
Statistics</H3>
<PRE>
protected <B>Statistics</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="setMethod(spie.Statistics.Method)"><!-- --></A><H3>
setMethod</H3>
<PRE>
public static void <B>setMethod</B>(<A HREF="../spie/Statistics.Method.html" title="enum in spie">Statistics.Method</A>&nbsp;i)</PRE>
<DL>
<DD>Sets the method to use for calculating weights.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>i</CODE> - the new method</DL>
</DD>
</DL>
<HR>

<A NAME="setTotalNumberOfReads(int)"><!-- --></A><H3>
setTotalNumberOfReads</H3>
<PRE>
public static void <B>setTotalNumberOfReads</B>(int&nbsp;i)</PRE>
<DL>
<DD>Sets the total number of reads.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>i</CODE> - the new total number of reads</DL>
</DD>
</DL>
<HR>

<A NAME="setReadLength(int)"><!-- --></A><H3>
setReadLength</H3>
<PRE>
public static void <B>setReadLength</B>(int&nbsp;i)</PRE>
<DL>
<DD>Sets the short read length.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>i</CODE> - the new short read length</DL>
</DD>
</DL>
<HR>

<A NAME="getWeightOfExon(gffParser.Exon, java.util.ArrayList, java.lang.Boolean)"><!-- --></A><H3>
getWeightOfExon</H3>
<PRE>
public static double <B>getWeightOfExon</B>(<A HREF="../gffParser/Exon.html" title="class in gffParser">Exon</A>&nbsp;exon,
                                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleMRNAReads,
                                     <A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Boolean.html" title="class or interface in java.lang">Boolean</A>&nbsp;isEndExon)</PRE>
<DL>
<DD>Gets the weight of an exon.
 This weight is dependent on what method is used.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>compatibleMRNAReads</CODE> - the ArrayList of compatible reads for the MRNA<DD><CODE>exon</CODE> - the exon<DD><CODE>isEndExon</CODE> - the end exon
<DT><B>Returns:</B><DD>the weight</DL>
</DD>
</DL>
<HR>

<A NAME="getAllReads_Per_KTotalPossiblePositions(int, int, java.util.ArrayList, boolean)"><!-- --></A><H3>
getAllReads_Per_KTotalPossiblePositions</H3>
<PRE>
private static double <B>getAllReads_Per_KTotalPossiblePositions</B>(int&nbsp;absoluteStart,
                                                              int&nbsp;absoluteEnd,
                                                              <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads,
                                                              boolean&nbsp;endExon)</PRE>
<DL>
<DD>Gets the ratio of all reads (junction and body) to the total number of possible start
 positions for a short read.
 
 The total number of positions can be calculated using the equation
 [L-2(v-1)]+[(r-1)-2(v-1)] (for inner exons)
 [L-2(v-1)] (for end exons)
 
 where
 L is the length of the exon
 r is the length of the short read
 v is the length of the overlap
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the absolute start of the exon<DD><CODE>absoluteEnd</CODE> - the absolute end of the exon<DD><CODE>compatibleShortReads</CODE> - the ArrayList of compatible ShortReads<DD><CODE>endExon</CODE> - the boolean that specifies whether the current exon is an end exon
<DT><B>Returns:</B><DD>the ratio of all reads : total number of start positions</DL>
</DD>
</DL>
<HR>

<A NAME="getAverage_ReadsPerBase_PerExon(int, int, java.util.ArrayList)"><!-- --></A><H3>
getAverage_ReadsPerBase_PerExon</H3>
<PRE>
public static double <B>getAverage_ReadsPerBase_PerExon</B>(int&nbsp;absoluteStart,
                                                     int&nbsp;absoluteEnd,
                                                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</PRE>
<DL>
<DD>Gets the ratio of the sum of reads per base to the total number of bases.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the absolute start of the exon<DD><CODE>absoluteEnd</CODE> - the absolute end of the exon<DD><CODE>compatibleShortReads</CODE> - the ArrayList of compatible ShortReads
<DT><B>Returns:</B><DD>the sum of reads body per base divided by total number of bases</DL>
</DD>
</DL>
<HR>

<A NAME="getBodyReads_per_KExon(int, int, java.util.ArrayList)"><!-- --></A><H3>
getBodyReads_per_KExon</H3>
<PRE>
public static double <B>getBodyReads_per_KExon</B>(int&nbsp;absoluteStart,
                                            int&nbsp;absoluteEnd,
                                            <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</PRE>
<DL>
<DD>Gets the number of body short reads divided by the exon length.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the absolute start of an exon<DD><CODE>absoluteEnd</CODE> - the absolute end of an exon<DD><CODE>compatibleShortReads</CODE> - the compatible short reads
<DT><B>Returns:</B><DD>the number of reads for the exon divided by the length of the exon</DL>
</DD>
</DL>
<HR>

<A NAME="getStandardDeviation(int, int, java.util.ArrayList)"><!-- --></A><H3>
getStandardDeviation</H3>
<PRE>
public static double <B>getStandardDeviation</B>(int&nbsp;absoluteStart,
                                          int&nbsp;absoluteEnd,
                                          <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getStandardDeviation_ReadsPerBase(int, int, java.util.ArrayList)"><!-- --></A><H3>
getStandardDeviation_ReadsPerBase</H3>
<PRE>
public static double <B>getStandardDeviation_ReadsPerBase</B>(int&nbsp;absoluteStart,
                                                       int&nbsp;absoluteEnd,
                                                       <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;compatibleShortReads)</PRE>
<DL>
<DD>Gets the standard deviation_ reads per base.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the absolute start<DD><CODE>absoluteEnd</CODE> - the absolute end<DD><CODE>compatibleShortReads</CODE> - the compatible short reads
<DT><B>Returns:</B><DD>the standard deviation_ reads per base</DL>
</DD>
</DL>
<HR>

<A NAME="getCompatibleReads(gffParser.MRNA, java.util.ArrayList)"><!-- --></A><H3>
getCompatibleReads</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt; <B>getCompatibleReads</B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                                                 <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</PRE>
<DL>
<DD>Gets a ArrayList of compatible ShortReads.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>isoform</CODE> - the isoform you want to find compatible reads for<DD><CODE>iSAMRecords</CODE> - the arrayList of SAMRecords you want to check (Non-Compatible)
<DT><B>Returns:</B><DD>An ArrayList of compatible ShortReads</DL>
</DD>
</DL>
<HR>

<A NAME="getCompatibleSamRecords(gffParser.MRNA, java.util.ArrayList)"><!-- --></A><H3>
getCompatibleSamRecords</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt; <B>getCompatibleSamRecords</B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                                                                           <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSamRecords)</PRE>
<DL>
<DD>Gets a compatible ArrayList of SAMRecords
 (it calls getCompatibleShortReadsFirst).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>isoform</CODE> - the input isoform<DD><CODE>iSamRecords</CODE> - the input array of SAMRecords (NonCompatible ShortReads)
<DT><B>Returns:</B><DD>the ArrayList of compatible SAMRecords</DL>
</DD>
</DL>
<HR>

<A NAME="convertShortReadsToSamRecords(java.util.ArrayList)"><!-- --></A><H3>
convertShortReadsToSamRecords</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt; <B>convertShortReadsToSamRecords</B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Read.html" title="class in spie">Read</A>&gt;&nbsp;iShortReads)</PRE>
<DL>
<DD>Convert short reads to sam records. Does not check for compatibility
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>iShortReads</CODE> - the input array of ShortReads
<DT><B>Returns:</B><DD>the ArrayList of SamRecords</DL>
</DD>
</DL>
<HR>

<A NAME="getShortReadIfCompatible(net.sf.samtools.SAMRecord, java.util.ArrayList, java.util.Collection)"><!-- --></A><H3>
getShortReadIfCompatible</H3>
<PRE>
private static <A HREF="../spie/Read.html" title="class in spie">Read</A> <B>getShortReadIfCompatible</B>(net.sf.samtools.SAMRecord&nbsp;samRecord,
                                             <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt;&nbsp;samInterval,
                                             <A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../gffParser/Exon.html" title="class in gffParser">Exon</A>&gt;&nbsp;exonList)</PRE>
<DL>
<DD>Gets a new ShortRead if the SAMRecord is compatible
 
 1) First find the exon where the short read begins
 2) Catch the case of a single interval --See if the short read is completely within the exon
 3) If more than one interval, check each interval with the next ones. Make sure short read
 interval has 'start of exon'  and end of short read has 'end of exon' (except the last)
 4) Check that the last interval matches
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>samRecord</CODE> - the SAMRecord that you want to find out is compatible or not<DD><CODE>samInterval</CODE> - the ArrayList of Intervals that contain the genomic coordinates
 for which a short read spans<DD><CODE>exonList</CODE> - the Collection of exons that the SAMRecord might be found in or cross
<DT><B>Returns:</B><DD>a new ShortRead is returned if it is compatible. If it is not, null is returned.</DL>
</DD>
</DL>
<HR>

<A NAME="getIsoformDensityMap(int, int, gffParser.MRNA, java.util.ArrayList)"><!-- --></A><H3>
getIsoformDensityMap</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>,<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt; <B>getIsoformDensityMap</B>(int&nbsp;absoluteStart,
                                                            int&nbsp;absoluteEnd,
                                                            <A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                                                            <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSamRecords)</PRE>
<DL>
<DD>Gets the isoform density map.A density map is a hash map that has all the genomic coordinates as
 keys and the number of short reads that cross it as values
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the start of the gene<DD><CODE>absoluteEnd</CODE> - the end of the gene<DD><CODE>isoform</CODE> - the isoform for which you want to get a density map of<DD><CODE>iSamRecords</CODE> - the ArrayList of SAMRecords from which a density map will be generated from
<DT><B>Returns:</B><DD>The HashMap of <Integer,Integer> where the key is the genomic coordinate and the value
 is the number of short reads that cross it. This hash map is generate only using compatible reads</DL>
</DD>
</DL>
<HR>

<A NAME="getDensityMap(int, int, java.util.ArrayList)"><!-- --></A><H3>
getDensityMap</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/HashMap.html" title="class or interface in java.util">HashMap</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>,<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt; <B>getDensityMap</B>(int&nbsp;absoluteStart,
                                                     int&nbsp;absoluteEnd,
                                                     <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</PRE>
<DL>
<DD>Gets the density map. A density map is a hash map that has all the specified coordinates as
 keys and the count of short reads that cross it as values
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>absoluteStart</CODE> - the start of the gene<DD><CODE>absoluteEnd</CODE> - the end of the gene<DD><CODE>iSAMRecords</CODE> - the ArrayList of SAMRecords from which a density map will be generated from
<DT><B>Returns:</B><DD>The HashMap of <Integer,Integer> where the key is the genomic coordinate and the value
 is the number of short reads that cross it</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfBodyReads(gffParser.MRNA, int, int, java.util.ArrayList)"><!-- --></A><H3>
getNumberOfBodyReads</H3>
<PRE>
public static int <B>getNumberOfBodyReads</B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                                       int&nbsp;absoluteStart,
                                       int&nbsp;absoluteEnd,
                                       <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</PRE>
<DL>
<DD>Gets the number of body reads as specified by the coordinates.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>isoform</CODE> - the isoform from which you want the compatible reads<DD><CODE>absoluteStart</CODE> - the absolute start of the exon<DD><CODE>absoluteEnd</CODE> - the absolute end of the exon<DD><CODE>iSAMRecords</CODE> - the Array of SAMRecords (Non-Compatible)
<DT><B>Returns:</B><DD>the number of body reads in the specified region</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfJunctionReads(gffParser.MRNA, int, int, java.util.ArrayList)"><!-- --></A><H3>
getNumberOfJunctionReads</H3>
<PRE>
public static int <B>getNumberOfJunctionReads</B>(<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&nbsp;isoform,
                                           int&nbsp;absoluteStart,
                                           int&nbsp;absoluteEnd,
                                           <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;iSAMRecords)</PRE>
<DL>
<DD>Gets the number of junction reads as specified by the coordinates.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>isoform</CODE> - the isoform from which you want the compatible reads<DD><CODE>absoluteStart</CODE> - the absolute start of the exon<DD><CODE>absoluteEnd</CODE> - the absolute end of the exon<DD><CODE>iSAMRecords</CODE> - the Array of SAMRecords (Non-Compatible)
<DT><B>Returns:</B><DD>the number of junction reads in the specified region</DL>
</DD>
</DL>
<HR>

<A NAME="setOverhang(int)"><!-- --></A><H3>
setOverhang</H3>
<PRE>
public static void <B>setOverhang</B>(int&nbsp;iOverhang)</PRE>
<DL>
<DD>Sets the overhang.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>iOverhang</CODE> - the new overhang</DL>
</DD>
</DL>
<HR>

<A NAME="getWeightOfJunction(drawableObjects.Junction)"><!-- --></A><H3>
getWeightOfJunction</H3>
<PRE>
public static double <B>getWeightOfJunction</B>(<A HREF="../drawableObjects/Junction.html" title="class in drawableObjects">Junction</A>&nbsp;junction)</PRE>
<DL>
<DD>Gets the weight of junction.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>junction</CODE> - the junction
<DT><B>Returns:</B><DD>the weight of junction</DL>
</DD>
</DL>
<HR>

<A NAME="getRPKM(java.util.ArrayList, java.util.ArrayList, int)"><!-- --></A><H3>
getRPKM</H3>
<PRE>
public static double <B>getRPKM</B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;net.sf.samtools.SAMRecord&gt;&nbsp;allSAMRecord,
                             <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt;&nbsp;constitutiveIntervals,
                             int&nbsp;totalNumberOfReads)</PRE>
<DL>
<DD>Gets the rPKM.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>allSAMRecord</CODE> - the all sam record<DD><CODE>constitutiveIntervals</CODE> - the constitutive intervals<DD><CODE>totalNumberOfReads</CODE> - the total number of reads
<DT><B>Returns:</B><DD>the rPKM</DL>
</DD>
</DL>
<HR>

<A NAME="getConstitutiveIntervals(java.util.ArrayList)"><!-- --></A><H3>
getConstitutiveIntervals</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="../spie/Interval.html" title="class in spie">Interval</A>&gt; <B>getConstitutiveIntervals</B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt;&nbsp;inputArrayOfConstitutivePositions)</PRE>
<DL>
<DD>Gets the constitutive intervals.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>inputArrayOfConstitutivePositions</CODE> - the input array of constitutive positions
<DT><B>Returns:</B><DD>the constitutive intervals</DL>
</DD>
</DL>
<HR>

<A NAME="getConstituitiveBases(java.util.Collection)"><!-- --></A><H3>
getConstituitiveBases</H3>
<PRE>
public static <A HREF="http://java.sun.com/javase/6/docs/api/java/util/ArrayList.html" title="class or interface in java.util">ArrayList</A>&lt;<A HREF="http://java.sun.com/javase/6/docs/api/java/lang/Integer.html" title="class or interface in java.lang">Integer</A>&gt; <B>getConstituitiveBases</B>(<A HREF="http://java.sun.com/javase/6/docs/api/java/util/Collection.html" title="class or interface in java.util">Collection</A>&lt;<A HREF="../gffParser/MRNA.html" title="class in gffParser">MRNA</A>&gt;&nbsp;mrnaList)</PRE>
<DL>
<DD>Returns an ArrayList of genomic coordinates which are constitutive.
 
 This method works by plotting all the coordinates for each MRNA. It then iteratively
 goes through each coordinate and determines if the number of "hits" for the coordinate
 equal the number of MRNAs in the list
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>mrnaList</CODE> - is a Collection of MRNA.
<DT><B>Returns:</B><DD>An ArrayList of constitutive genomic coordinates are returned.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Statistics.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../spie/Read.html" title="class in spie"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../spie/Statistics.Method.html" title="enum in spie"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../index.html?spie/Statistics.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Statistics.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
